transport/conntable.go

```diff
@@ -4,27 +4,73 @@ import (
 	"time"
 
 	"github.com/stefankopieczek/gossip/log"
+	"github.com/stefankopieczek/gossip/timing"
 )
 
 // Fields of connTable should only be modified by the dedicated goroutine called by Init().
 // All other callers should use connTable's associated public methods to access it.
 type connTable struct {
-	conns   map[string]*connWatcher
-	stopped bool
+	conns        map[string]*connWatcher
+	connRequests chan *connRequest
+	updates      chan *connUpdate
+	expiries     chan string
+	stop         chan bool
+	stopped      bool
 }
 
 type connWatcher struct {
-	addr   string
-	conn   *connection
-	timer  *time.Timer
-	update chan *connection
-	stop   chan bool
+	addr       string
+	conn       *connection
+	timer      timing.Timer
+	updated    chan bool
+	expiryTime time.Time
+	expiry     chan<- string
+	stop       chan bool
 }
 
 // Create a new connection table.
 func (t *connTable) Init() {
-	log.Info("Init conntable %p")
+	log.Info("Init conntable %p", t)
 	t.conns = make(map[string]*connWatcher)
+	t.connRequests = make(chan *connRequest)
+	t.updates = make(chan *connUpdate)
+	t.expiries = make(chan string)
+	t.stop = make(chan bool)
+	go t.manage()
+}
+
+// Management loop for the connTable.
+// Handles notifications of connection updates, expiries of connections, and
+// the termination of the routine.
+func (t *connTable) manage() {
+	for {
+		select {
+		case request := <-t.connRequests:
+			watcher := t.conns[request.addr]
+			if watcher != nil {
+				request.responseChan <- watcher.conn
+			} else {
+				request.responseChan <- nil
+			}
+		case update := <-t.updates:
+			t.handleUpdate(update)
+		case addr := <-t.expiries:
+			if t.conns[addr].expiryTime.Before(time.Now()) {
```

Comment:
I think I disagree. What about the following scenario?
* We push socket S1 for address A.
* [Time elapses]
* While the ConnTable routine is busy doing something else, the timer expires and tries to push an expiry to the expiry chan. It blocks.
* Meanwhile, a higher-level caller gives us a new socket, S2, for A. It blocks on the update chan.
* The ConnTable goroutine spins around, and hits select { } again. It randomly chooses the update chan and pushes S2 to the conntable.
* The routine spins around again, hits select { } again, and finally pushes our expiry. We process it, and wrongly expire S2. 

In fact, we should ignore spurious expiries of this kind. I don't think there's any scenario where we'll fail this 'if' other than because of a previously-expires socket.

I might be wrong in the above analysis. Thoughts?